plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java

```diff
@@ -0,0 +1,340 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.openlineage;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Inject;
+import io.airlift.log.Logger;
+import io.openlineage.client.OpenLineage;
+import io.openlineage.client.OpenLineage.ColumnLineageDatasetFacetBuilder;
+import io.openlineage.client.OpenLineage.InputDataset;
+import io.openlineage.client.OpenLineage.JobBuilder;
+import io.openlineage.client.OpenLineage.OutputDataset;
+import io.openlineage.client.OpenLineage.RunEvent;
+import io.openlineage.client.OpenLineage.RunFacet;
+import io.openlineage.client.OpenLineage.RunFacetsBuilder;
+import io.openlineage.client.OpenLineageClient;
+import io.trino.plugin.openlineage.config.OpenLineageListenerConfig;
+import io.trino.spi.eventlistener.EventListener;
+import io.trino.spi.eventlistener.OutputColumnMetadata;
+import io.trino.spi.eventlistener.QueryCompletedEvent;
+import io.trino.spi.eventlistener.QueryContext;
+import io.trino.spi.eventlistener.QueryCreatedEvent;
+import io.trino.spi.eventlistener.QueryFailureInfo;
+import io.trino.spi.eventlistener.QueryIOMetadata;
+import io.trino.spi.eventlistener.QueryMetadata;
+import io.trino.spi.eventlistener.QueryOutputMetadata;
+import io.trino.spi.eventlistener.QueryStatistics;
+import io.trino.spi.resourcegroups.QueryType;
+
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.String.format;
+
+public class OpenLineageListener
+        implements EventListener
+{
+    private final OpenLineage openLineage = new OpenLineage(URI.create("https://github.com/trinodb/trino/plugin/trino-openlineage"));
+    private final OpenLineageClient client;
+    private final String namespace;
+    private final Set<QueryType> includeQueryTypes;
+    private final ObjectMapper queryStatisticsMapper;
+    private static final Logger logger = Logger.get(OpenLineageListener.class);
+
+    @Inject
+    public OpenLineageListener(OpenLineageClient client, OpenLineageListenerConfig listenerConfig)
+            throws Exception
+    {
+        this.client = client;
+        this.namespace = listenerConfig.getTrinoURI().toString().replace(listenerConfig.getTrinoURI().getScheme(), "trino");
+
+        this.includeQueryTypes = listenerConfig.getIncludeQueryTypes();
+
+        this.queryStatisticsMapper = createQueryStatisticsObjectMapper();
```

Comment:
is there anything particularly not ok with this approach? for JsonCodec I think I would also need to bring another dependency.

Reply:
`JsoncCodec` dependency is fine, it is from `io.airlift:json` - Plus `JsonCodec` unifies the list of modules required for `ObjectMapper` so we don't want to install them once again and it would be nice if we could have a unified version of Json ser/deser

Reply:
ok, I can add it but would need some help understanding this - mapJsonCodec enables me to serialize a class to json string or the other way around (it has methods fromJson and toJson) but that's not what I need, that's why i use directly object mapper - to get a property and put it in a hashmap (required by RunFacet.putAll())